### ISSUE TITLE
About simulation layer

### ISSUE THREAD

**Issue by lifeofstudy** [rmin2000/WaDiff#6]
I recently re-read your paper and learned a lot. I have a few questions. I read your code for pre-training watermark decoders, that is, the code train.py in the file stegastamp. I found that you designed a relatively complex training strategy when training, including the design of L2_loss_weight. Could you tell me the motivation or thinking process of this design? Secondly, you said in the paper: "similar to the simulation layer in [11], we transform w(x) with random data augmentations during the training stage to improve the robustness of the watermark system", but I don't seem to find the corresponding code. Looking forward to your early reply, thank you.

**Comment by rmin2000** [comment]
Hi, thanks for your question.

The code for StegaStamp training is adapted from [WatermarkDM](https://github.com/yunqing-me/WatermarkDM/blob/main/string2img/train_cifar10.py), where they use the L2_loss_weight to adaptively adjust the weight parameter for quality control. Basically, during the initial training stage, the L2_loss_weight is 0, where the network focuses only on bit matching. When the bit acc exceeds a certain threshold, L2_loss_weight is increased, and the network gradually learns the capability to preserve the image quality. I just find this could help accelerate the training process, nevertheless, you can still use a single static parameter to balance the BCE and image-quality loss as described in this paper.

Why I adopted this design and where the simulation layer is: Well, I rewrote all the code after my original was unintentionally wiped from the server. I tried to remember/reproduce all the necessary components in my paper, but there might be a little difference compared to my original implementation. And thanks for letting me know. I will implement the simulation layer and the code for the diffusion part as soon as possible. Besides, you can find the data augmentation in (https://github.com/facebookresearch/stable_signature/blob/48261686883ba86f533836ad47c35040afcfe37b/finetune_ldm_decoder.py#L326); my implementation is very similar to this.
